### PR TITLE
Server dev

### PR DESCRIPTION
**Pull request by evansiroky**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -8,13 +10,15 @@ notifications:
 node_js:
   - '6'
 before_install:
-  - npm i -g yarn
-  - npm i -g codecov
+  - npm i -g yarn codecov
 before_script:
   - npm prune
+  - sleep 15
```
Comment (trevorgerhardt) [inline .travis.yml]: What is the `sleep 15` for?

File: .travis.yml [hunk 2]
```diff
@@ -8,13 +10,15 @@ notifications:
...
+  - mongo mydb_test --eval 'db.addUser("travis", "test");'
```
Comment (trevorgerhardt) [inline .travis.yml]: Yikes, I assume you found this elsewhere as a recommended approach? Seems like there would be a better way to do this.

File: .travis.yml [hunk 3]
```diff
@@ -8,13 +10,15 @@ notifications:
 node_js:
   - '6'
 before_install:
-  - npm i -g yarn
-  - npm i -g codecov
+  - npm i -g yarn codecov
 before_script:
   - npm prune
```
Comment (trevorgerhardt) [inline .travis.yml]: Prune shouldn't be needed when installing with `yarn`

File: __tests__/client/actions/agency.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
+/* global describe */
+
+import {makeGenericModelActionsTests} from '../../test-utils/actions'
+import {blankAgency} from '../../test-utils/mock-data'
+
+import * as agency from '../../../client/actions/agency'
+
+describe('actions > agency', () => {
+  makeGenericModelActionsTests({
```
Comment (trevorgerhardt) [inline __tests__/client/actions/agency.js:9]: A few thoughts...this is a lot of configuration to automate generating tests. I would advise against this level of abstraction. Having repeated describe/it/expect statements for each action isn't a bad thing. Once you need to add more detailed tests you can just expand upon those simpler ones. Like the component snapshot tests in SE. We could automate generating a snapshot for each component, but having the test there allows us to expand upon the test when needed.

File: __tests__/client/components/breadcrumb-bar.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    '/organization/organization-2/group/create',
+    '/group/group-2',
+    '/group/group-2/add',
+    '/group/group-2/commuter/create',
+    '/commuter/commuter-2/edit',
+    '/organization/organization-2/site/create',
+    '/site/site-2/edit'
+  ]
+
+  testCases.forEach((testPath) => {
```
Comment (trevorgerhardt) [inline __tests__/client/components/breadcrumb-bar.js:33]: I would prefer iterating over the array inside the `it` statement here and having multiple `expect` clauses within.

File: __tests__/client/containers/add-commuters.js [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+    // mount component
+    const tree = mount(
+      <Provider store={mockStore}>
+        <AddCommuters
+          params={{organizationId: 'organization-1'}}
+          />
+      </Provider>
+    )
+    expect(pretty(tree.find(AddCommuters).html())).toMatchSnapshot()
```
Comment (trevorgerhardt) [inline __tests__/client/containers/add-commuters.js:37]: Beautifying HTML for better snapshot rendering? Is that necessary?

File: __tests__/client/containers/edit-organization.js [hunk 7]
```diff
@@ -1,24 +1,139 @@
...
+    )
+
+    // give each text field some input
+    tree.find('input').map((input) => input.simulate('change', {target: {value: 'My new value'}}))
+
+    // submit form
+    tree.find('form').find('button').simulate('click')
+
+    // expect create action
+    organizationExpectations.expectCreateAction({
```
Comment (trevorgerhardt) [inline __tests__/client/containers/edit-organization.js:66]: This seems like really a powerful testing utility but it might be too powerful ;)

File: __tests__/client/reducers/commuter.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+import {handleActions} from 'redux-actions'
+
+import {commuterSal, mockCommuter, mockGroupCreation, mockStores} from '../../test-utils/mock-data'
+import {makeGenericReducerTestCases} from '../../test-utils/reducers'
+
+import * as commuter from '../../../client/reducers/commuter'
+
+describe('client > reducers > commuter', () => {
+  makeGenericReducerTestCases({
```
Comment (trevorgerhardt) [inline __tests__/client/reducers/commuter.js:11]: These abstractions definitely seem like overkill...

File: __tests__/server/organization.js [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    agencyId: mongoose.Types.ObjectId(),
+    name: 'test-org'
+  }
+
+  makeRestEndpointTests({
+    endpoints: {
+      'Collection GET': {},
+      'Collection POST': {
+        creationData: initOrganizationData,
+        customAssertions: (json) => {
```
Comment (trevorgerhardt) [inline __tests__/server/organization.js:24]: Seeing this here emphasizes the point for me. If you need configuration objects this big along with a place to insert custom assertions it's a pretty clear sign to me that the abstraction should not have been made.

File: __tests__/test-utils/actions.js [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+
+const deepEqual = (actual, expected) => {
+  if (Array.isArray(actual)) {
+    for (let i = 0; i < actual.length; i++) {
+      deepEqual(actual[i], expected[i])
+    }
+  } else if (actual === null) {
+    expect(actual).toEqual(expected)
+  } else if (typeof actual === 'object') {
+    Object.keys(actual).forEach((k) => {
```
Comment (trevorgerhardt) [inline __tests__/test-utils/actions.js:11]: Any reason we're not using `toMatchObject` here? https://facebook.github.io/jest/docs/api.html#tomatchobjectobject

File: __tests__/test-utils/actions.js [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+      deepEqual(actual[k], expected[k])
+    })
+  } else {
+    expect(actual).toEqual(expected)
+  }
+}
+
+const expectFetchActionAndGetNextFn = (action, url, optionsToAssertEqual) => {
```
Comment (trevorgerhardt) [inline __tests__/test-utils/actions.js]: We should still use the traditional `function functionName () {}` style for functions of any significant length that don't use the benefits of the arrow function.

File: client/components/add-commuters.js [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+      createCommuter(commutersToCreate)
+    } else {
+      const newGroup = {...this.state}
+      newGroup.commuters = commutersToCreate
+      delete newGroup.newCommuters
+      createGroup(newGroup)
+    }
+  }
+
+  makeCommuterTable (commuters) {
```
Comment (trevorgerhardt) [inline client/components/add-commuters.js]: I'd recommend using the simple component pattern for things like this:

```js
function CommuterTable ({commuters}) {
  return (
    <BootstrapTable data={commuters}>
    ...
    </BootstrapTable>
  )
}
```

File: client/components/add-commuters.js [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+            <form>
+              <FieldGroup
+                label='Name'
+                name='name'
+                onChange={this.handleChange}
+                placeholder='Enter name'
+                type='text'
+                value={groupName}
+                />
+              <Dropzone
```
Comment (trevorgerhardt) [inline client/components/add-commuters.js:110]: Oi! Drag and drop is definitely a "nice to have" in my book. What's the motivation for including this in the first version instead of just a simple file upload?

File: client/components/agencies.js [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+    // props
+    agencies: PropTypes.array.isRequired
+  }
+
+  componentWillMount () {
+    this.props.loadAgencies()
+  }
+
+  _toolsRenderer = (cell, row) => {
+    const doDelete = () => { this.props.deleteAgency(row) }
```
Comment (trevorgerhardt) [inline client/components/agencies.js]: Since these functions are created on render they will always evaluate to false when comparing the props, thus requiring a re-render (assuming you are doing a deep equal on `shouldComponentUpdate`. I'd recommend pulling them out into class methods.

File: client/components/analysis/histogram.js [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+
+    // props
+    analysis: PropTypes.object.isRequired,
+    groupName: PropTypes.string.isRequired,
+    siteName: PropTypes.string.isRequired
+  }
+
+  constructor (props) {
+    super(props)
+    this.state = {
```
Comment (trevorgerhardt) [inline client/components/analysis/histogram.js]: This can be done right on the class, see this [example](https://github.com/conveyal/react-select-geocoder/blob/master/index.js#L28)

File: client/components/analysis/histogram.js [hunk 16]
```diff
@@ -0,0 +1,262 @@
...
+    // calculate with selected metric
+    this._calculateSeries({...this.state})
+  }
+
+  /**
+   * Calculate series values given selected metric
+   *
+   * @param  {Object} newState The new state to work with
+   */
+  _calculateSeries = (newState) => {
```
Comment (trevorgerhardt) [inline client/components/analysis/histogram.js]: This style should only be used if you are passing the function to be called elsewhere (like in an event handler) and it needs to be bound to the instance.

File: client/components/analysis/individuals.js [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+
+  constructor (props) {
+    super(props)
+    this.state = {
+      position: settings.map.focus,
+      zoom: 8
+    }
+  }
+
+  _costRenderer = (cell, row) => {
```
Comment (trevorgerhardt) [inline client/components/analysis/individuals.js]: Functions that don't act on class properties and therefore don't require being attached to an instance should live outside of the class on their own.

Comment (trevorgerhardt) [reply client/components/analysis/individuals.js]: Simple rule: no `this` then no need to be in the class.

File: client/components/breadcrumb-bar.js [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+    // props
+    agency: PropTypes.object.isRequired,
+    analysis: PropTypes.object.isRequired,
+    commuter: PropTypes.object.isRequired,
+    group: PropTypes.object.isRequired,
+    organization: PropTypes.object.isRequired,
+    site: PropTypes.object.isRequired
+  }
+
+  _makeNavItems () {
```
Comment (trevorgerhardt) [inline client/components/breadcrumb-bar.js:16]: There has gotta be an easier/more sustainable way to do this...

File: client/utils/components.js [hunk 19]
```diff
@@ -0,0 +1,53 @@
+import {settings} from './env'
+
+/**
+ * Calculate number of elements in array less than or equal to target values
+ *
+ * @param  {Number[]} arr     A sorted array of numbers
+ * @param  {[Number]} target  The number to be less than or equal to
+ * @return {int}              The number of elements in the array less than or equal to the target value
+ */
+export function calcNumLessThan (arr, target) {
```
Comment (trevorgerhardt) [inline client/utils/components.js:10]: Is this being calculated on an array long enough that `arr.filter((v) => v <= target).length` is too slow?

File: configurations/default/settings.yml [hunk 20]
```diff
@@ -0,0 +1,42 @@
+entries:
+  - client/index.js:assets/index.js
+  - client/index.css:assets/index.css
+map:
```
Comment (trevorgerhardt) [inline configurations/default/settings.yml]: This type of data should go in the `store.yml`: https://github.com/conveyal/scenario-editor/blob/dev/configurations/default/store.yml

File: configurations/test/messages.yml [hunk 21]
```diff
@@ -0,0 +1,22 @@
+agency:
```
Comment (trevorgerhardt) [inline configurations/test/messages.yml]: If a configuration file is missing from the config you are pointing to it will try to retrieve that file from the default configuration. Duplicate messages and store files will usually be unnecessary.

File: server/models/agency.js [hunk 22]
```diff
@@ -0,0 +1,18 @@
...
+const agencySchema = new Schema({
+  name: {
+    required: true,
+    type: String
+  }
+})
+
+agencySchema.pre('remove', function (next) {
+  // CASCADE DELETE
+  Organization.remove({ agencyId: this._id }).exec()
```
Comment (trevorgerhardt) [inline server/models/agency.js]: We should switch all removes to soft deletes. Have a flag across all models that signifies whether or not it's been "deleted". The way I implemented it in Modeify was here: https://github.com/conveyal/modeify/blob/dev/lib/plugins/mongoose-trash.js

File: server/routes/index.js [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+const makeGetModelResponse = (childModels, res, isCollection) => (err, data) => {
+  const genericResponder = makeGenericModelResponse(res)
+  if (!Array.isArray(data)) {
+    data = [data]
+  }
+
+  if (!childModels) return genericResponder(err, isCollection ? data : data[0])
+
+  // wow, awesome, with normalized mongoose models I get to make extra queries
+  // to the db to do joins!  </sarcasm>
```
Comment (trevorgerhardt) [inline server/routes/index.js]: These less than 20 lines of code must have been very painful for you to write :)

Also, you probably want to use [populate](http://mongoosejs.com/docs/populate.html)

**Review by trevorgerhardt — COMMENTED** [review]
My biggest reaction is that I would highly recommend reevaluating your approach to abstracting away boilerplate. This project is very young. The requirements are going to change over and over again. These large configuration objects for generating the rest endpints, tests, reducers, etc. are going to get even more complex. There will be more and more edge cases. Don't fear the boilerplate code and un-DRY code in these instances. There's a good chance you'll need to remove the abstraction in the future.

**Comment by evansiroky** [comment]
Here are my points of motivation for abstracting the REST endpoints:
- The REST pattern of GET, POST, PUT, DELETE is very common and understood
- When I was writing boilerplate, I found myself being very inconsistent in everything which was confusing me in various places.  By using an abstraction the code is very similar everywhere.
- By using the abstraction future changes should be able to be implemented more quickly since there is a template to do so.

File: .travis.yml (hunk 3 above)
Comment (evansiroky) [reply .travis.yml]: Yes, it isn't needed to do an install.  I guess I'm just being paranoid in thinking what if someone ran this command, wouldn't I want it to still work?  I'll remove it.

File: .travis.yml (hunk 1 above)
Comment (evansiroky) [reply .travis.yml]: Copied from [here](https://docs.travis-ci.com/user/database-setup/#MongoDB-does-not-immediately-accept-connections).  May not be needed since I reorganized tests.  I'll try removing it.

File: __tests__/client/actions/agency.js (hunk 4 above)
Comment (evansiroky) [reply __tests__/client/actions/agency.js:9]: For something as common as REST endpoints, I like this configuration because it is consistent and much less code to screw up.  Furthermore, it is entirely possible to only configure some but not all of the rest endpoints.

File: __tests__/client/components/breadcrumb-bar.js (hunk 5 above)
Comment (evansiroky) [reply __tests__/client/components/breadcrumb-bar.js:33]: If the test was written that way, the single test case would abort on the first failure and not show any further failures.

File: __tests__/client/containers/add-commuters.js (hunk 6 above)
Comment (evansiroky) [reply __tests__/client/containers/add-commuters.js:37]: Yes, otherwise it's a big mess of a single string and hard to examine differences.  I'm using html snapshots here because the react-bootstrap-table generates massive snapshots with enzyme-to-json.

File: __tests__/client/containers/edit-organization.js (hunk 7 above)
Comment (evansiroky) [reply __tests__/client/containers/edit-organization.js:66]: There are tradeoffs to re-using the actions in the store.  The container tests have sort of become an integration test of container, component and action all at once.  If one were to send in mock functions and test the component directly that way it would truly become a unit test.  However, I do like the integration-style test since I can make sure all these things work together.

File: __tests__/client/reducers/commuter.js (hunk 8 above)
Comment (evansiroky) [reply __tests__/client/reducers/commuter.js:11]: At least they're consistently used across the app.

File: __tests__/server/organization.js (hunk 9 above)
Comment (evansiroky) [reply __tests__/server/organization.js:24]: The configuration and custom assertions here are small compared to the boilerplate code that doesn't need to be written as a result.  And they force consistency across the app.

File: __tests__/test-utils/actions.js (hunk 10 above)
Comment (evansiroky) [reply __tests__/test-utils/actions.js:11]: That isn't quite the same as deep-equal.

File: __tests__/test-utils/actions.js (hunk 11 above)
Comment (evansiroky) [reply __tests__/test-utils/actions.js]: agreed, will replace

File: client/components/add-commuters.js (hunk 12 above)
Comment (evansiroky) [reply client/components/add-commuters.js]: ok

File: client/components/add-commuters.js (hunk 13 above)
Comment (evansiroky) [reply client/components/add-commuters.js:110]: It works with minimal extra configuration (I added some css and that's it).

File: client/components/agencies.js (hunk 14 above)
Comment (evansiroky) [reply client/components/agencies.js]: ok

File: client/components/analysis/histogram.js (hunk 15 above)
Comment (evansiroky) [reply client/components/analysis/histogram.js]: ok

File: client/components/analysis/histogram.js (hunk 16 above)
Comment (evansiroky) [reply client/components/analysis/histogram.js]: ok

File: client/components/analysis/individuals.js (hunk 17 above)
Comment (evansiroky) [reply client/components/analysis/individuals.js]: ok

File: client/components/breadcrumb-bar.js (hunk 18 above)
Comment (evansiroky) [reply client/components/breadcrumb-bar.js:16]: I'm open to suggestions.  :)

File: client/utils/components.js (hunk 19 above)
Comment (evansiroky) [reply client/utils/components.js:10]: well, binary search is O(log(n)) vs O(n) for the filter method, so binary search is always going to be faster in the worst case for array length > 3.  This method is used primarily in the possibilities analysis componenet.  That component has a slider which can call this function a lot, so performance is important here.

File: configurations/default/settings.yml (hunk 20 above)
Comment (evansiroky) [reply configurations/default/settings.yml]: Were you thinking this kind of data could be put into a user store and a user table in the database?

File: configurations/test/messages.yml (hunk 21 above)
Comment (evansiroky) [reply configurations/test/messages.yml]: ok

Comment (evansiroky) [reply configurations/test/messages.yml]: It may affect snapshotting in tests tho.

File: server/models/agency.js (hunk 22 above)
Comment (evansiroky) [reply server/models/agency.js]: What is the advantage of preserving the data?

File: server/routes/index.js (hunk 23 above)
Comment (evansiroky) [reply server/routes/index.js]: Yes, they were painful to write.  And it'll be slower since it's more roundtrips to the database.  And there's no referential integrity.

If I use `populate` on arrays of children ids then I have to refactor all of the models that have children models.  I'd have to do even more manual work of adding the ids of new models to their parent and then deleting them from their parent's children array upon deletion.  Seems like more work that way.

File: __tests__/client/containers/add-commuters.js (hunk 6 above)
Comment (trevorgerhardt) [reply __tests__/client/containers/add-commuters.js:37]: 👍

File: __tests__/client/components/breadcrumb-bar.js (hunk 5 above)
Comment (trevorgerhardt) [reply __tests__/client/components/breadcrumb-bar.js:33]: Good call then!

File: __tests__/test-utils/actions.js (hunk 10 above)
Comment (trevorgerhardt) [reply __tests__/test-utils/actions.js:11]: I'm surprised there isn't a built in one to Jest. Seems like a lot of people would want this assertion.

File: client/components/add-commuters.js (hunk 13 above)
Comment (trevorgerhardt) [reply client/components/add-commuters.js:110]: It's adding complexity without adding a feature. Those should be avoided really as much as possible with this project.

File: client/utils/components.js (hunk 19 above)
Comment (trevorgerhardt) [reply client/utils/components.js:10]: Just trying to watch out for instances of premature optimization. The simplest working solution is usually the best working solution.

File: configurations/default/settings.yml (hunk 20 above)
Comment (trevorgerhardt) [reply configurations/default/settings.yml]: Nope. Front-end state is not the same as DB store. Application state isn't always serialized and stored. store.yml is used specifically with redux in mind.

File: client/components/add-commuters.js [hunk 24]
```diff
@@ -161,3 +150,14 @@ export default class AddCommuters extends Component {
     )
   }
 }
+
+function makeCommuterTable (commuters) {
```
Comment (trevorgerhardt) [inline client/components/add-commuters.js]: Ah, meant that this could be made into a react component. `<CommuterTable commuters={this.state.newCommuters} />`

**Comment by trevorgerhardt** [comment]
It's not creating an abstraction itself that is the problem, but the size of these abstractions. The boilerplate is replaced by large enough configuration objects that I would refer to them  as a [DSL](https://en.wikipedia.org/wiki/Domain-specific_language). Many smaller abstractions that require less configuration and could be pieced together would be much more ideal. Twenty different ten line functions are sometimes more understandable then one two-hundred line function.

**Comment by trevorgerhardt** [comment]
By the way, I don't think we should change the instances I'm referring too right now. If it works, keep it. What I would ask for is to keep the abstractions smaller going forward. It really comes down to those configuration objects. If they're that big, it's a clear sign that multiple smaller abstractions should be made. Piece them together. It will make the code more reusable and understandable.
